### ISSUE TITLE
`./mach test-tidy-faster`

### ISSUE THREAD

**Issue by Wafflespeanut** [servo/servo#9088]
Currently, tidy is slow (see #8969), almost entirely because we run the WPT lint tool. So, @Manishearth just threw an idea that we could have a mach command `test-tidy-faster` which ignores the `check_wpt_lint_errors` function in tidy.

Code: `python/servo/testing_commands.py` and `python/tidy.py`

**Comment by SimonSapin** [comment]
Or can we fix the slowness? Profiling the lint tool shows that it spends 73% of it’s time parsing HTML.

```
python2 -m cProfile -o /tmp/profile lint
gprof2dot.py -f pstats /tmp/profile | xdot
```

I’ve tried changing the wpt lint tool to use lxml.html instead of html5lib. It causes some errors (below), but it runs in 9 seconds instead of 38 on my machine. If the errors can be resolved, we could change the tool to use lxml when available and fall back on html5lib otherwise. (lxml can be a pain to compile/install, so it probably shouldn’t be required.) @jgraham, what do you think?

With that change, 41% of the time is now spent in `fnmatch` so there may be more optimization opportunity. 

```
PARSE-FAILED: Unable to parse file dom/nodes/Document-createElement-namespace-tests/empty.html
PARSE-FAILED: Unable to parse file html/browsers/browsing-the-web/navigating-across-documents/blank.html
PARSE-FAILED: Unable to parse file html/browsers/the-window-object/apis-for-creating-and-navigating-browsing-contexts-by-name/close_script_defer-1.html
PARSE-FAILED: Unable to parse file html/dom/dynamic-markup-insertion/opening-the-input-stream/004-1.html
PARSE-FAILED: Unable to parse file webmessaging/without-ports/019-1.html
There were 5 errors (PARSE-FAILED: 5)

You must fix all errors; for details on how to fix them, see
https://github.com/w3c/web-platform-tests/blob/master/docs/lint-tool.md

However, instead of fixing a particular error, it's sometimes
OK to add a line to the lint.whitelist file in the root of the
web-platform-tests directory to make the lint tool ignore it.

For example, to make the lint tool ignore all 'PARSE-FAILED'
errors in the webmessaging/without-ports/019-1.html file,
you could add the following line to the lint.whitelist file.

PARSE-FAILED:webmessaging/without-ports/019-1.html

**Comment by Ms2ger** [comment]
For a web standards test suite, we should probably use an html parser that follows the spec. I wonder if we could hook up h5e...

**Comment by SimonSapin** [comment]
I’ve played a bit with https://github.com/SimonSapin/html5ever-python, but it needs more work

**Comment by SimonSapin** [comment]
Or we could have a Rust version of the lint program? I haven’t looked how big it is.

**Comment by jgraham** [comment]
I agree that using lxml is a no-go, unfortunately. On the other hand, a proper Python frontend for Kuchiki or the html5ever treebuilder would be something I would bite your hand off for.

Rewriting the tool in Rust would probably not lead to something that you could upstream, unfortunately (that is, I am not comfortable requiring other web-platform-tests users to use Rust components), so it would miss newly added lints (which is a thing that happens sometimes).

**Comment by Wafflespeanut** [comment]
FWIW, this is just a temporary workaround, so that we could just run the faster version locally and let Travis take care of the entire check. The slowness of tidy is being discussed in #8969.

**Comment by jgraham** [comment]
One option would be to make the wpt lint able to take a list of files to lint. Then you would get the file list from the changes to the wpt directory since the last sync point.

**Comment by dsprenkels** [comment]
I would like to take this issue.

My idea is to add an option to `./mach test-tidy` (`--changes`) so that it will only test changes in the working directory from the last merge by @bors-servo.

**Comment by Wafflespeanut** [comment]
You got it!

**Comment by Wafflespeanut** [comment]
Not exactly fixed by #9118, still has a bug, but that's covered by #9253
